designs/statefulset-disruption.md

```diff
@@ -0,0 +1,739 @@
+
+**Pros:**
+- Leaves decision to each cloud/storage provider 
+- Can opt-in to this behavior for specific CSI Drivers (Perhaps via Helm parameter)
+- Only delays termination of nodes with stateful workloads.
+**Cons:**
+- Delays node termination and finalizer deletion by a worst-case of 20 seconds. (We can skip waiting on the volumes of non-drainable pods to make the average case lower)
+- Other CSI Drivers must opt-in
+
+### Alternatives Considered
+
+**Implement B1 in `kubernetes-sigs/karpenter`**
+
+Instead of solving this inside the [`c.cloudProvider.Delete`](https://github.com/aws/karpenter-provider-aws/blob/9cef47b6df77ec9e0a39dc6f4a4ecd1aab504ae3/pkg/cloudprovider/cloudprovider.go#L179), solve this inside [termination controller's reconciler loop](https://github.com/kubernetes-sigs/karpenter/blob/94d5b41d4711b1c21fc0264f1b29db6a64b95caf/pkg/controllers/node/termination/controller.go#L77) (as is done in [today's proof-of-concept implementation](https://github.com/kubernetes-sigs/karpenter/pull/1294))
+
+**Pros:**
+- Karpenter-provider-aws does not need to know about Kubernetes volume lifecycle
+**Cons:**
+- [Open Question] EBS may be the only storage provider where detaching volumes before terminating the instance matters. If this is the case, the delay before instance termination is not worth it for customers of other storage/cloud providers.
+- Should not hardcode cloud-provider specific CSI Drivers in upstream project. 
+
+**Karpenter Polls EC2 for volume detachments before calling EC2 TerminateInstance**
+
+Karpenter-provider-aws can poll EC2 DescribeVolumes before making an EC2 TerminateInstance call. 
+
+**Pros:**
+- Solves issue
+**Cons:**
+- If volumes are not unmounted due to problem A, volumes cannot be detached anyway. Termination is the fastest way forward.
+- Karpenter has to worry about EBS volume lifecycle. 
+
+## Appendix 
+
+### Z. Document TODOs
+
+- Mention that volumes with Multi-Attach enabled do not have this problem
+- Move images to gists before PR merge
+- Increase resolution of pngs. 
+- Expand Appendix terminology + timeline sections
+- Prove what potential data corruption issues we are talking about. 
+
+### A. Further Reading
+
+- [EBS CSI: What causes 6-minute delays in attaching volumes?](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#6-minute-delays-in-attaching-volumes)
+- Kubernetes:
+  - [Node Shutdowns](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/)
+  - [API-initiated Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/#how-api-initiated-eviction-works)
+
+
+### B. Terminology
+Daemonset: Pod scheduled on every Node. 
+EBS: Elastic Block Store. An Amazon Web Service that 
+Kubelet: 
+StatefulSet: Manages the deployment and scaling of a set of Pods, and provides guarantees about the ordering and uniqueness of these Pods. These uniqueness gauruntees are valuable when your workload needs persistent storage. 
```

Comment:
needs and extra line or double space at the end of the previous line, to properly break line in MD

Reply:
or just bullets